Improve clarity of `core::sync::atomic` docs about "Considerations" in regards to CAS operations

fu5ha opened the pull request:
## Motivation

The existing documentation for atomic `fetch_update` (and other similar methods) has a section that reads like so:

> ### Considerations
> This method is not magic; it is not provided by the hardware. It is implemented in
> terms of `AtomicBlah::compare_exchange_weak`, and suffers from the same drawbacks.
> In particular, this method will not circumvent the [ABA Problem].
>
> [ABA Problem]: https://en.wikipedia.org/wiki/ABA_problem

The wording here seems to imply that the drawbacks being discusses are caused by the *`weak` version* of `compare_exchange`, and that one may avoid those drawbacks by using `compare_exchange` instead. Indeed, a conversation in the `#dark-arts` channel on the Rust community discord based on this interpretation led to this PR.

In reality, the drawbacks are inherent to implementing such an operation based on *any* compare-and-swap style operation, as opposed to an [LL,SC](https://en.wikipedia.org/wiki/Load-link/store-conditional) operation, and they apply equally to `compare_exchange` and `compare_exchange_weak` as well.

## Changes

- Rewords existing Considerations section on `fetch_update` and friends to make clear that the limitations are inherent to an implementation based on any CAS operation, rather than the weak version of `compare_exchange` in particular. New version:

> ### Considerations
>
> This method is not magic; it is not provided by the hardware, and does not act like a
> critical section or mutex.
>
> It is implemented on top of an atomic [compare-and-swap operation], and thus is subject to
> the usual drawbacks of CAS operations. In particular, be careful of the [ABA problem]
> if this atomic integer is an index or more generally if knowledge of only the *bitwise value*
> of the atomic is not in and of itself sufficient to ensure any required preconditions.
>
> [ABA Problem]: https://en.wikipedia.org/wiki/ABA_problem
> [compare-and-swap operation]: https://en.wikipedia.org/wiki/Compare-and-swap

- Add Considerations to `compare_exchange` and `compare_exchange_weak` which details similar considerations and when they may be relevant. New version:

> ### Considerations
>
> `compare_exchange` is a [compare-and-swap operation] and thus exhibits the usual downsides
> of CAS operations. In particular, a load of the value followed by a successful
> `compare_exchange` with the previous load *does not ensure* that other threads have not
> changed the value in the interim. This is usually important when the *equality* check in
> the `compare_exchange` is being used to check the *identity* of a value, but equality
> does not necessarily imply identity. In this case, `compare_exchange` can lead to the
> [ABA problem].
>
> [ABA Problem]: https://en.wikipedia.org/wiki/ABA_problem
> [compare-and-swap operation]: https://en.wikipedia.org/wiki/Compare-and-swap

ibraheemdev commented on library/core/src/sync/atomic.rs:
```diff
@@ -1271,11 +1297,16 @@ impl AtomicBool {
+    /// [`compare_exchange`][Self::compare_exchange]\*, which is a [CAS operation], and thus
+    /// suffers from the usual drawbacks of CAS operations, in particular the [ABA problem].
+    ///
+    /// \* It is actually implemented using [`compare_exchange_weak`][Self::compare_exchange_weak],
```
I don't think the strong vs. weak CAS is observable behavior for `fetch_update`. Given that I'm inclined to either remove this note or just mention `compare_exchange_weak` in the previous paragraph, now that it clarifies that the drawbacks are linked to the use of CAS, not whether or not the CAS is weak or not.

fu5ha committed:
core docs: improve clarity of considerations about atomic CAS operations

- Rewords existing Considerations section on `fetch_update` and friends
to make clear that the limitations are inherent to an implementation based on any
CAS operation, rather than the weak version of `compare_exchange` in particular
- Add Considerations to `compare_exchange` and `compare_exchange_weak`
which details similar considerations and when they may be relevant.

ibraheemdev commented:
Thanks. @bors r+

bors commented:
:pushpin: Commit 6c55fbacbb84911f87abc12647de0804f6177898 has been approved by `ibraheemdev`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ibraheemdev 6c55fbacbb84911f87abc12647de0804f6177898 -->
<!-- homu: {"type":"Approved","sha":"6c55fbacbb84911f87abc12647de0804f6177898","approver":"ibraheemdev","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:pushpin: Commit 2095211fb674df018ddd37d37677a1d2eaf61e49 has been approved by `ibraheemdev`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ibraheemdev 2095211fb674df018ddd37d37677a1d2eaf61e49 -->
<!-- homu: {"type":"Approved","sha":"2095211fb674df018ddd37d37677a1d2eaf61e49","approver":"ibraheemdev","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
